### ISSUE TITLE
Springfox with Spring 4.x annotations. Interfaces having Spring4.x annotations not supported in springfox

### ISSUE THREAD

**Issue by sac10nikam** [springfox/springfox#1389]
Note: Resteasy 2.3.5 supports interfaces for swagger documentation. But springfox doesn't

**Problem Statement:**
I am trying to generate Swagger Specification for SpringMVC Rest project. Springfox is not supporting annotations (Spring 4.x annotations) in interfaces for generating swagger documentation.

**Key jars:**
springfox-swagger-ui
springfox-swagger2
swagger-springmvc

**My environment:**

<details>
    <summary>**pom.xml**</summary>
```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>springmvc.swagger</groupId>
	<artifactId>swagger</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<description>Swaggerpoc</description>

	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>1.3.0.RELEASE</version>
	</parent>

	<properties>
		<java-version>1.8</java-version>
		<api-common-version>1.0.3-SNAPSHOT</api-common-version>
		<api-messaging-version>1.0.3-SNAPSHOT</api-messaging-version>
	</properties>

	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
			<groupId>io.springfox</groupId>
			<artifactId>springfox-swagger-ui</artifactId>
			<version>2.5.0</version>
		</dependency>

		<!-- https://mvnrepository.com/artifact/io.springfox/springfox-swagger2 -->
		<dependency>
			<groupId>io.springfox</groupId>
			<artifactId>springfox-swagger2</artifactId>
			<version>2.5.0</version>
		</dependency>
		<!-- 
		<dependency>
			<groupId>io.springfox</groupId>
			<artifactId>springfox-core</artifactId>
			<scope>system</scope>
			<version>1</version>
			<systemPath>C:/Users/sachin_nikam/Desktop/springfox-core.jar</systemPath>
		</dependency> -->
	

		<!-- https://mvnrepository.com/artifact/com.mangofactory/swagger-springmvc -->
		<dependency>
			<groupId>com.mangofactory</groupId>
			<artifactId>swagger-springmvc</artifactId>
			<version>1.0.2</version>
		</dependency>

		<dependency>
			<groupId>org.springframework</groupId>
			<artifactId>spring-core</artifactId>
			<version>4.2.5.RELEASE</version>
		</dependency>
	</dependencies>

	<build>
		<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>
		</plugins>
	</build>

	<repositories>
		<repository>
			<id>spring-releases</id>
			<url>https://repo.spring.io/libs-release</url>
		</repository>
	</repositories>
	<pluginRepositories>
		<pluginRepository>
			<id>spring-releases</id>
			<url>https://repo.spring.io/libs-release</url>
		</pluginRepository>
	</pluginRepositories>
</project>
```
</details>

<details>
    <summary>**Interface**</summary>
```java
import Response;
import Contact;

@RequestMapping("/v1/contact")
@Api(value = "Contact services")
public interface ContactController {
    /**
     * Purpose: Fetch contacts.
     * 
     * 
     * @param employeeId
     * @param companyId
     * 
     * @return ReturnResponse
     * @throws Exception
     * @see success data
     * 
     */
	@ApiOperation(value = "getContacts", notes = "Gets the list of contacts", responseContainer = "List<Contact>")
    @ApiResponses(value = { @ApiResponse(code = 200, message = "Success", response = Response.class),
            @ApiResponse(code = 401, message = "Unauthorized", response =Response.class),
            @ApiResponse(code = 403, message = "Forbidden", response = Response.class),
            @ApiResponse(code = 404, message = "Not Found", response = Response.class),
            @ApiResponse(code = 500, message = "Internal server error occured", response = Response.class) })
    @RequestMapping(value = "/{companyId}/{employeeId}/contacts", method = RequestMethod.GET)
    Response<List<Contact>> getContacts(@PathVariable String employeeId,
            @PathVariable String companyId);
}
```
</details>

**Models**:

<details>
    <summary>**Reponse.java**</summary>
```java
@JsonAutoDetect
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Response<T> implements Serializable {
    private static final long serialVersionUID = 1L;
    private T data;
    private Error error;
    private String statusCode;
    private String statusText;
   
    public T getData() {
        return this.data;
    }

    @JsonProperty("_error")
    public Error getError() {
        return this.error;
    }

    @JsonProperty("_statusCode")
    public String getStatusCode() {
        return this.statusCode;
    }

    @JsonProperty("_statusText")
    public String getStatusText() {
        return this.statusText;
    }
    public void setData(T data) {
        this.data = data;
    }
    public void setError(Error error) {
        this.error = error;
    } 

    public void setStatusCode(String statusCode) {
        this.statusCode = statusCode;
    }

    public void setStatusText(String statusText) {
        this.statusText = statusText;
    }
}
```
</details>

<details>
    <summary>**Contact.java**</summary>
```java


import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;

import java.io.Serializable;

import com.fasterxml.jackson.annotation.JsonAutoDetect;

@ApiModel
@JsonAutoDetect
public class Contact implements Serializable {

    private static final long serialVersionUID = 1L;

    
    @ApiModelProperty(notes = "The contact address 1", required = true)
    private String address1;
    @ApiModelProperty(notes = "The contact address 2", required = false)
    private String address2;

    // Getters and setters
    /**
     * @return the address1
     */
    public String getAddress1() {
        return address1;
    }
    /**
     * @param address1 the address1 to set
     */
    public void setAddress1(String address1) {
        this.address1 = address1;
    }
    /**
     * @return the address2
     */
    public String getAddress2() {
        return address2;
    }
    /**
     * @param address2 the address2 to set
     */
    public void setAddress2(String address2) {
        this.address2 = address2;
    }
}
```
</details>

**Controller Implementation:**
<details>
    <summary>**Contact Controller Impl**</summary>
```java
import Contact;
import Response;

@RestController
public class ContactControllerImpl implements ContactController  {

   
    /**
     * Purpose: Fetches Contacts details.
     * 
     * 
     * @param employeeId
     * @param companyId
     * @return Response
     * 
     * @throws Exception
     * @see success data
     * 
     */
	 @Override
    public Response<List<Contact>> getContacts(@PathVariable String employeeId,
            @PathVariable String companyId) {
       final List<Contact> contactlist = new ArrayList<Contact>();
        return contactlist;
    }
}
```

Please run this code at your end. Interfaces are not supported by springfox API.

http://host:port/swagger-ui.html should list me interfaces instead of controller impl

**Comment by dilipkrish** [comment]
@sac10nikam Looks like your controller doesn't implement the interface.

```java
public class ContactControllerImpl /* missing implements ContactController */ {
//...
}
```

**Comment by sac10nikam** [comment]
Thanks for pointing out. But I had missed it to add. Can you please run this as a project and see. The swagger doesn't work for above code if put swagger annotations and spring 4.x annotations in interface and in controller impl class I implement this interface. And override its method.

Ideally it should work. Please help.

**Comment by dilipkrish** [comment]
Please share a github repo that reproduces the issue.

**Comment by sac10nikam** [comment]
ok. Sure Will upload it in next 30 mins. Thanks.

**Comment by sac10nikam** [comment]
Hi Dilip,

I have uploaded sample maven project.  https://github.com/sac10nikam/swagger-poc
Please have a look at it. I am online. We can talk/chat/communicate here.

Many Thanks,
Sachin

**Comment by sac10nikam** [comment]
Hi,

I have uploaded my sample project. Please have a look.

Thanks,
Sachin

**Comment by dilipkrish** [comment]
See interfaces where? It seems to be working just fine.

**Comment by sac10nikam** [comment]
In the attached image we could see controller implementation instead of
interface. Can you please check image from last message.
I.e. HTTP://host:port/context/swagger-springmvc.html should show list of
interfaces instead of controller.

It is fine but I want to support interface having swagger annotations and
spring 4.x annotations.

Thanks,
Sachin

**Comment by dilipkrish** [comment]
Those are just the descriptions. You can control those via annotations in the interface. In your case using the `@Api#tags`

**Comment by dilipkrish** [comment]
Also just so it's clear, interface and implementations don't matter in the swagger Api description. It's the url, media type, query parameters etc. everything else can be controlled via annotations.

**Comment by sac10nikam** [comment]
Thanks for your quick reply. 
Please do below given change in project and see the output. This issue I want to resolve.

Change config/Application.java
Change @ComponentScan(basePackageClasses = { ContactControllerImpl.class })
to @ComponentScan(basePackageClasses = { ContactController.class })

and build it and run the application and hit http://localhost:8080/swagger-ui.html

Thanks,
Sachin

**Comment by dilipkrish** [comment]
I don't see any difference. What are you expecting to resolve?

If you change the Api annotation like below you'll see what you're expecting to see.

```java
@RequestMapping("/v1/contact")
@Api(value = "Contact services", tags = "Contact Services", description = "Contact Services")
//Instead of @Api(value = "Contact services")
public interface ContactController {
//...
}
```

**Comment by sac10nikam** [comment]
Thanks.. I guess it is working.. I had checked it in the morning.. There was no change required. But I am still wondering why it was not show the list of interfaces.

Anyway thanks a lot for your time. Have a gr8 day.

**Comment by sermojohn** [comment]
Hi, I have a similar issue to the one reported here. I am using  springfox:2.6.1 with springboot:1.5.2.RELEASE and I would like to generate swagger documentation from java interfaces using spring mvc annotations. However my REST controller interfaces seem to be ignored.

Any help is welcome.

**Comment by dilipkrish** [comment]
@sermojohn this should be fixed in 2.7

**Comment by apissdks** [comment]
+1 

@dilipkrish does it mean that just providing springfox with the RESTController interfaces we should be able to generate the docs/spec ?
